### PR TITLE
Feat/14 make view model and repository for food items

### LINKED ISSUE #14: Make view model and repository for food items
Fully implement View Model and repository for Food Items, such that food items can be added to firebase

### PR DESCRIPTION
**Pull request by alextheprophecy**
# Add FoodItem Data Models and Repository Integration

## **Summary**
This PR introduces the following key updates to the data models and repository pattern used for managing food items in the application:

1. Added a new `FoodItem.kt` file that meets the parameter requirements discussed during the **07.10 scrum meeting**.
2. Implemented new repository components to support data retrieval and management:
   - `FoodItemRepository`
   - `FoodItemRepositoryFirestore`
   - `ListFoodItemsViewModel`

### Changes in **FoodItem.kt**:
- Introduced new data classes including `FoodItem`, `Quantity`, `FoodLocation`, and `NutritionFacts` to support robust and structured data storage.
- Added enum classes: `FoodCategory`, `FoodStatus`, `_Location`, and `FoodUnit` to better categorize and manage the food items.

### **FoodItemRepository** & **FoodItemRepositoryFirestore**:
- Implemented similarly to Bootcamp _TodosRepository_
- Added a repository interface (`FoodItemRepository`) for CRUD (create, read, update, delete) operations.
- Implemented a Firestore-backed repository (`FoodItemRepositoryFirestore`) to interact with the Firebase Firestore database.
- Methods include adding, updating, retrieving, and deleting food items, with error handling to manage potential issues

### **ListFoodItemsViewModel**:
- Introduced a new `ViewModel` class that manages the UI-related data for displaying food items.
- Integrated live data flows using `StateFlow` to ensure efficient state management and updates for the UI.
- Added error handling and logic to manage selected items and view updates.

## **Testing**
- **Tested the `FoodItemRepositoryFirestore` class:**
  - Created unit tests using `RobolectricTestRunner` to simulate Android environment and Firebase Firestore interactions.
  - Verified the `getNewUid()` method to ensure it generates the correct unique identifier for a `FoodItem`.
  - Added a test for the `getFoodItems()` method to confirm that it successfully retrieves food items from the mocked Firestore collection.
  - Implemented a test for the `addFoodItem()` method to validate that the Firestore `collection()` method is properly called when adding a new `FoodItem`.
  - Tested the `deleteFoodItemById()` method to verify that the correct document reference is being deleted from Firestore.
  
> **Note:** Tests were performed using mocked instances of Firestore components to ensure they interact as expected without requiring a live database connection.

## **Follow-Up Tasks**
- Improve the user experience when displaying error messages in case of failed operations.
- Add unit tests to cover edge cases for data conversion and repository methods.

### CODE REVIEW

**Review by Dearop — APPROVED** [review]
Good job, needs some tweaks to fit into the scope of the application. Issues will be resolved in a future merge.

File: app/src/main/java/com/android/shelfLife/model/foodItem/FoodItem.kt
Comment (david-schroeter) [inline app/src/main/java/com/android/shelfLife/model/foodItem/FoodItem.kt:1]: Too many inlined comments in this file simply describe what each line is doing, which adds unnecessary weight to the code without providing valuable insights. Inline comments should focus on explaining complex logic, business rules, or decisions that are not immediately clear from the code itself. Removing redundant comments will make the code easier to read and maintain.

File: app/src/main/java/com/android/shelfLife/model/foodItem/FoodItemRepositoryFirestore.kt [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+import android.util.Log
+import com.google.firebase.Timestamp
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+
+class FoodItemRepositoryFirestore(private val db: FirebaseFirestore) : FoodItemRepository {
+
+    private val collectionPath = "foodItems"
```
Comment (david-schroeter) [inline app/src/main/java/com/android/shelfLife/model/foodItem/FoodItemRepositoryFirestore.kt:11]: It is a good practice to declare constant values in a companion object. This ensures that only one instance of the constant is created, reducing memory usage and avoiding unnecessary resource consumption.

```
companion object {
    private const val COLLECTION_PATH = "foodItems"
}   
```

File: app/src/main/java/com/android/shelfLife/model/foodItem/FoodItemRepositoryFirestore.kt
Comment (david-schroeter) [inline app/src/main/java/com/android/shelfLife/model/foodItem/FoodItemRepositoryFirestore.kt:1]: Great implementation of the `FoodItemRepository`! It looks clean and well-organized. To provide an alternative approach and showcase a more idiomatic Kotlin style, I would like to suggest a different way of writing the `getFoodItem` function using coroutines.

Kotlin coroutines allow you to write asynchronous code in a way that looks and behaves like regular, sequential code. They help to manage non-blocking operations such as I/O tasks without the need for callbacks, improving readability and reducing the complexity of asynchronous programming.

If you're not already familiar with coroutines, the [official documentation](https://kotlinlang.org/docs/coroutines-overview.html) is a great place to start.

---
Below is the `getFoodItem` function as an example of how coroutines can be used. This approach simplifies the code by avoiding callbacks, making it easier to read and manage asynchronous operations.

```
suspend fun getFoodItems2(): List<FoodItem> {
    return try {
        db.collection(collectionPath).get().await().mapNotNull {
            Log.e("FoodItemRespository", "Doc: $it")
            convertToFoodItem(it)
        }
    } catch (e: Exception) {
        Log.e("FoodItemRepository", "Error fetching food items", e)
        emptyList()
        // Handle error how you would normally without need for callbacks, can also
        // throw custom exception that can be caught by the caller of `getFoodItem2`
    }
}
```
> **Note**: While getFoodItem is used as an example here, the same coroutine-based approach can be applied to other functions as well, allowing you to avoid callbacks throughout the repository. By rewriting each function this way, you can make the codebase more consistent, cleaner, and easier to maintain.

That said, this is just a suggestion, and it’s entirely up to the team to decide what approach you prefer. I’m providing this as an alternative way to write asynchronous code so that the team is aware of the option, and can choose what best suits the project’s needs.

File: app/src/main/java/com/android/shelfLife/model/foodItem/ListFoodItemsViewModel.kt
Comment (david-schroeter) [inline app/src/main/java/com/android/shelfLife/model/foodItem/ListFoodItemsViewModel.kt:1]: Great implementation of the `ViewModel`! To continue with my example of using coroutines instead of callbacks, you could simplify your `getFoodItems` function like this:

```
fun getFoodItems() {
    try {
        viewModelScope.launch {
            _foodItems.value = repository.getFoodItems2()
        }
    } catch (e: Exception) {
        _onFail(e)
    }
}
```

This makes the function cleaner, eliminates the need for callbacks, and leverages the lifecycle-aware nature of `viewModelScope`. You can apply this pattern to other functions as well to streamline your code.

**Review by david-schroeter — COMMENTED** [review]
Great job on this PR! 

It adds valuable functionality to the app, and the code is clean and well-written. The PR description is awesome and provides all the necessary context to understand what’s going on, making it easy to follow the changes.

I especially like the implementation of the CRUD repository to manage FoodItem, and the tests for the repository are well done. It would be great to see some additional tests for the ViewModel to ensure full coverage.

I’ve also provided an example of using coroutines instead of callbacks for asynchronous programming. This is just a suggestion for a more Kotlin-idiomatic approach, feel free to adopt it if the team prefers. The main goal is to make sure you're aware of this option.

Keep up the great work!
